worlds/pokemon_emerald/__init__.py

```diff
@@ -558,33 +547,56 @@ def pre_fill(self) -> None:
-            hm_items.sort(key=lambda item: hm_priority.get(item.name, 0))
-
-            # Un-exclude HM locations, since we need to put progression items on them
-            for location in hm_locations:
-                location.progress_type = LocationProgressType.DEFAULT \
-                    if location.progress_type == LocationProgressType.EXCLUDED \
-                    else location.progress_type
-
-            collection_state = self.multiworld.get_all_state(False)
-
-            # In specific very constrained conditions, fill_restrictive may run
-            # out of swaps before it finds a valid solution if it gets unlucky.
-            # This is a band-aid until fill/swap can reliably find those solutions.
-            attempts_remaining = 2
-            while attempts_remaining > 0:
-                attempts_remaining -= 1
-                self.random.shuffle(hm_locations)
-                try:
-                    fill_restrictive(self.multiworld, collection_state, hm_locations, hm_items,
-                                     single_player_placement=True, lock=True, allow_excluded=True)
-                    break
-                except FillError as exc:
-                    if attempts_remaining == 0:
-                        raise exc
-
-                    logging.debug(f"Failed to shuffle HMs for player {self.player}. Retrying.")
-                    continue
+            hm_items.sort(key=lambda item: hm_priority.get(item.name, 0), reverse=True)
+
+            # Build state
+            # Badges are either in the item pool, or already placed and collected during sweep
+            state = CollectionState(self.multiworld, True)
+            for item in my_progression_items:
+                state.collect(item, True)
+            state.sweep_for_advancements(my_locations)
+
+            # Shuffle HMs
+            self.fill_subset_with_retries(hm_items, hm_locations, state)
+
+    def fill_subset_with_retries(self, items: list[PokemonEmeraldItem], locations: list[PokemonEmeraldLocation], state: CollectionState):
+        # Un-exclude locations, since we need to put progression items on them
+        for location in locations:
+            location.progress_type = LocationProgressType.DEFAULT \
+                if location.progress_type == LocationProgressType.EXCLUDED \
+                else location.progress_type
+
+        # In specific very constrained conditions, `fill_restrictive` may run
+        # out of swaps before it finds a valid solution if it gets unlucky.
+        attempts_remaining = 2
+        while attempts_remaining > 0:
+            attempts_remaining -= 1
+            locations_copy = locations.copy()
+            items_copy = items.copy()
+            self.random.shuffle(locations_copy)
+            try:
+                fill_restrictive(self.multiworld, state, locations_copy, items_copy, single_player_placement=True,
+                                    lock=True, allow_excluded=True)
```

Comment:
You already forcibly un-exclude the locations, so there should be no need for this
```suggestion
                                    lock=True)
```